### PR TITLE
Fix top level await in middleware

### PR DESCRIPTION
**Pull request by khuezy**
### What?

If the middleware uses top level await, then the module itself becomes a Promise in the _ENTRIES mapping... so edgeFunction would resolve to <promise>.default, which is undefined.
At runtime, it errors: o is not a function

### Why?

Enables middleware to use top level await - for use cases where imported libraries in middleware.ts have top level await.
For example: loading auth secrets

### How?

Wraps the module loading with "await" so it properly awaits the potential `Promises`. If the module isn't a promise, then `await` does nothing.

Closes NEXT-
Fixes #

Discussion: https://github.com/vercel/next.js/discussions/48595

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 1c59645d6a18753adcf56335c4313ac69001e91e

`pnpm testheadless test/e2e/edge-compiler-module-exports-preference/index.test.ts` 
- Edge compiler module exports preference > favors the browser export

<details>
<summary>Expand output</summary>

  ● Edge compiler module exports preference › favors the browser export

    expect(received).toMatchObject(expected)

    - Expected  - 1
    + Received  + 9

      Object {
    -   "x-imported": "Browser",
    +   "cache-control": "no-store, must-revalidate",
    +   "connection": "close",
    +   "content-encoding": "gzip",
    +   "content-type": "text/html; charset=utf-8",
    +   "date": "Mon, 15 May 2023 16:45:01 GMT",
    +   "etag": "\"bujmi35eu9p9b\"",
    +   "transfer-encoding": "chunked",
    +   "vary": "Accept-Encoding",
    +   "x-powered-by": "Next.js",
      }

      52 |   it('favors the browser export', async () => {
      53 |     const response = await fetchViaHTTP(next.url, '/')
    > 54 |     expect(Object.fromEntries(response.headers)).toMatchObject({
         |                                                  ^
      55 |       'x-imported': 'Browser',
      56 |     })
      57 |   })

      at Object.<anonymous> (e2e/edge-compiler-module-exports-preference/index.test.ts:54:50)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by khuezy** [comment]
Test failing due to not setting the experimentl topLevelAwait :(

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 334abf1a37318a8b017a8451289fd6d975cff988

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

File: packages/next/src/server/web/sandbox/sandbox.ts [hunk 1]
```diff
@@ -89,8 +89,9 @@ export const run = withTaggedErrors(async function runWithTaggedErrors(params) {
 
   const edgeFunction: (args: {
     request: RequestData
-  }) => Promise<FetchEventResult> =
-    runtime.context._ENTRIES[`middleware_${params.name}`].default
+  }) => Promise<FetchEventResult> = (
+    await runtime.context._ENTRIES[`middleware_${params.name}`]
+  ).default
```
Comment (khuezy) [inline packages/next/src/server/web/sandbox/sandbox.ts:94]: If the middleware uses top level await, then the module itself becomes a `Promise` in the `_ENTRIES` mapping... so `edgeFunction` would resolve to `<promise>.default`, which is `undefined`.
At runtime, it errors: `o is not a function`

**Comment by socsieng** [comment]
@khuezy, looks like this change was applied in https://github.com/vercel/next.js/commit/212553958c671ea1f71d96fbc97ea4b9e5019bf3#diff-4fd4925f7835e9415a48410b8d0f7cb47acde9e8250372c085c9f948749411e0

Should this PR be closed?

**Comment by khuezy** [comment]
@socsieng yes. Thanks for the update, this PR was already addressed.

**Pull request closed without merging**